File: python/pyprepbuddy/utils/pivot_table.py
class PivotTable(object):
    """
    Represents a table of organized and summarized selected columns and rows of data.
    """
    def __init__(self, java_pivot_table):
        self.__table = java_pivot_table

    def add_entry(self, row_key, column_key, value):
        self.__table.addEntry(row_key, column_key, value)

    def value_at(self, row_key, column_key):
        return self.__table.valueAt(row_key, column_key)

File: src/main/scala/org/apache/datacommons/prepbuddy/utils/PivotTable.scala
package org.apache.datacommons.prepbuddy.utils

import scala.collection.mutable

/**
  * Represents a table of organized and summarized selected columns and rows of data.
  */
class PivotTable[T](defaultValue: T) extends Serializable {

    private var lookUpTable: mutable.Map[String, mutable.Map[String, T]] = {
        new mutable.HashMap[String, mutable.Map[String, T]]()
    }

    def transform[U](transformedFunction: (T) => U, defValue: U): PivotTable[U] = {
        val table = new PivotTable[U](defValue)
        for (rowTuple <- lookUpTable; columnTuple <- rowTuple._2)
            table.addEntry(rowTuple._1, columnTuple._1, transformedFunction(columnTuple._2))
        table
    }

    def addEntry(rowKey: String, columnKey: String, value: T): Unit = {
        if (!lookUpTable.contains(rowKey)) {
            val columnMap = new mutable.HashMap[String, T]().withDefaultValue(defaultValue)
            lookUpTable += (rowKey -> columnMap)
        }
        lookUpTable(rowKey) += (columnKey -> value)
    }

    def valueAt(rowKey: String, columnKey: String): T = {
        lookUpTable(rowKey)(columnKey)
    }
}
